File: Programmers/모의고사.js
// 맘에 안든다..
function solution(answers) {
    const one = [1, 2, 3, 4, 5]
    const two = [2, 1, 2, 3, 2, 4, 2, 5]
    const three = [3, 3, 1, 1, 2, 2, 4, 4, 5, 5]
    const grade = [0, 0, 0]

    for (let i=0; i<answers.length; i++) {
        if (answers[i] === one[i % 5]) grade[0] += 1
        if (answers[i] === two[i % 8]) grade[1] += 1
        if (answers[i] === three[i % 10]) grade[2] += 1
    }

    const max = grade.reduce((a, b) => Math.max(a, b))
    const answer = []
    grade.forEach((g, idx) => {
        if (g === max) answer.push(idx+1)
    })
    return answer;
}

File: Programmers/Level 1/모의고사.py
def solution(answers):
    answer = []
    grades = [0] * 3
    students = [
        [1, 2, 3, 4, 5],
        [2, 1, 2, 3, 2, 4, 2, 5],
        [3, 3, 1, 1, 2, 2, 4, 4, 5, 5]
    ]
    for idx, nums in enumerate(students):
        for i in range(len(answers)):
            if answers[i] == nums[i % len(nums)]:
                grades[idx] += 1
    _max = max(grades)
    for i, grade in enumerate(grades):
        if grade == _max:
            answer += [i+1]
    return answer


def solution2(answers):
    answer = []
    _max = 0
    students = {
        '1': [1, 2, 3, 4, 5],
        '2': [2, 1, 2, 3, 2, 4, 2, 5],
        '3': [3, 3, 1, 1, 2, 2, 4, 4, 5, 5]
    }
    for student, nums in students.items():
        cnt = 0
        for i in range(len(answers)):
            if answers[i] == nums[i % len(nums)]:
                cnt += 1
        answer += [(student, cnt)]
        if cnt > _max:
            _max = cnt
    answer = list(map(lambda y: int(y[0]), filter(lambda x: x[1] == _max, answer)))
    return answer


# print(solution([1, 2, 3, 4, 5]))  # [1]
# print(solution([1, 3, 2, 4, 2]))  # [1, 2, 3]
